### PR TITLE
Fix `/readyz` when upgrading cluster from 1.6 to 1.7

### PR DESCRIPTION
**Pull request by ffuugoo**
Fixes <https://github.com/qdrant/qdrant/issues/3193>.

When doing rolling upgrade from Qdrant v1.6.* to v1.7.*, the `/readyz` check would fail _indefinitely_, practically preventing graceful rolling upgrades in Kubernetes. 🤦‍♀️

The issue is that v1.6.* (and earlier) versions simply don't have gRPC APIs required for the `/readyz` check in v1.7.*, and "unimplemented" errors aren't handled properly.

This PR implements "unimplemented" errors handling during `/readyz`, allowing upgrade to v1.7.*.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by ffuugoo** [comment]
I'm testing this manually right now, will mark PR as ready if/once it works properly.

File: src/common/health.rs [hunk 1]
```diff
@@ -181,6 +181,8 @@ impl Task {
 
         let peer_address_by_id = self.consensus_state.peer_address_by_id();
 
+        let mut unimplemented_errors_count = 0;
```
Comment (timvisee) [inline src/common/health.rs:184]: Maybe we can put a comment before here, something like:

```suggestion
        // Endpoint we check is not available before Qdrant 1.7.0, allow checks on nodes with earlier versions to fail by counting them here
        let mut unimplemented_errors_count = 0;
```

**Review by timvisee — COMMENTED** [review]
Nice fix! I didn't test it myself yet, but it seems simple and effective.

File: src/common/health.rs [hunk 2]
```diff
@@ -181,6 +181,9 @@ impl Task {
 
         let peer_address_by_id = self.consensus_state.peer_address_by_id();
 
+        // Endpoint we check is not available before Qdrant 1.7.0, allow checks on nodes with earlier versions to fail by counting them here
```
Comment (agourlay) [inline src/common/health.rs:184]: Should we have a clear TODO to remove this trick in Qdrant 1.8.0?

Comment (timvisee) [reply src/common/health.rs:184]: We can do, but it doesn't hurt to keep it here for a bit longer.

**Comment by timvisee** [comment]
Rather than doing a simple subtraction on the required nodes when one returns not-implemented, I think doing this a little different will make it more reliable.

With the current approach, if you have 5 nodes and the first two return not-implemented, it shortcuts immediately assuming everything is okay. But if the other two nodes are up-to-date and return a wildly different cluster status, you have a problem.

Instead, if a node returns a not-implemented failure, I think we should decrease the list of nodes to consider and adjust all counters accordingly. In that case you make sure you query enough nodes running 1.7 to be confident everything is sound.

There is a downside, when updating the first node to 1.17, it will have to go to all nodes asking their status. This improves once more nodes reach 1.17 though, and completely disappears for further versions. But I think this is fine to do to ensure `/readyz` is more reliable.

I've partially discussed this with Roman already, but wanted to mention it publicly here as well.

**Review by generall — APPROVED** [review]

**Comment by timvisee** [comment]
I'll publish and release the patch version in about 10 minutes :+1:

**Pull request merged**